File: 0020-valid-parentheses/0020-valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    
    //Logic use stack push and pop method
    
//pseusodocode
    // Create a hashtable or object
    const hashTable = {
        '(':')',
        '{':'}',
        '[':']'    
    }
    // Initialize an empty stack
    
    let stack = []
    //Loop through an array
    if(s.length%2!==0)
        return false
    for(let i=0;i<s.length;i++)
    
   
        {
             //Push if the element is openning parentheses
            
     if(s[i]==='(' || s[i]=== '{' || s[i]==='[')
    {
        stack.push(s[i])
    }
          //if it is closing parenthesis pop 
           else if(s[i]===')' || s[i]=== '}' || s[i]===']')
            {
             //compare with the object or hashtable of itself
                let poped= hashTable[stack.pop()];
                if(poped !== s[i])
                    {
   
                        return false
                    }
                
                
            }
 
  
        }
    return !stack.length
 
    

    
};

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack =[]
        for val in s:
            if  val =='(' or  val == '{' or  val == '[' or  val == '[':
                stack.append(val)
            else:
                if not stack:
                    return False
                elif stack[-1] =='(' and  val == ')' or  stack[-1] =='[' and val == ']' or stack[-1] =='{' and  val == '}':
                    stack.pop()
                else:
                    return False
        return True if len(stack) == 0 else False        
                
                